src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -580,7 +698,166 @@ private static ImageHelper createImageProvider(JlinkConfiguration config,
                         targetPlatform.arch().byteOrder(), targetPlatform);
             }
         }
-        return new ImageHelper(cf, mods, targetPlatform, retainModulesPath, ignoreSigning);
+
+        // use the version of java.base module, if present, as
+        // the release version for multi-release JAR files
+        var version = cf.findModule("java.base")
+                        .map(ResolvedModule::reference)
+                        .map(ModuleReference::descriptor)
+                        .flatMap(ModuleDescriptor::version)
+                        .map(ModuleDescriptor.Version::toString)
+                        .map(Runtime.Version::parse)
+                        .orElse(Runtime.version());
+
+        Set<Archive> archives = mods.entrySet().stream()
+                .map(e -> newArchive(e.getKey(),
+                                     e.getValue(),
+                                     version,
+                                     ignoreSigning,
+                                     config))
+                .collect(Collectors.toSet());
+
+        return new ImageHelper(archives,
+                               targetPlatform,
+                               retainModulesPath,
+                               config.isGenerateRuntimeImage());
+    }
+
+    /**
+     * Linkable JDK runtimes support. Perform some sanity checks if we run
+     * jlink from the current run-time JDK image.
+     *
+     * @param log The log to write messages to.
+     * @param cf The current configuration
+     *
+     * @throws IOException
+     */
+    private static void sanityChecksLinkableJDKRuntime(PrintWriter log,
+                                                       Configuration cf,
+                                                       boolean verbose) throws IOException {
+        // Catch the case where we don't have a linkable JDK runtime. If so,
+        // we don't have the per module resource diffs in the modules image
+        try (InputStream inStream = getDiffInputStream()) {
+            if (inStream == null) {
```

Comment:
Suggest to define a method `isLinkableRuntime` in the new `LinkableRuntimeImage` class to test if it's a linkable runtime image.

Reply:
OK.